keps/sig-storage/20190125-online-growing-persistent-volume-size.md

```diff
@@ -0,0 +1,299 @@
+	return nil, nil
+```
+
+- If the file system resizing operation succeeds, `resizeFileSystem` will change `PVC.Status.Capacity` to the desired volume size in `PV.Spec.Capacity`, and remove the `PersistentVolumeClaimFileSystemResizePending` condition from `PVC.Status.Conditions`, then this resizing request is marked as finished.
+
+- If any operation fails inside `GenerateVolumeFSResizeFunc`, an event will be added to the according pod and an error will be returned to `Reconciler`, `Reconciler` will log this error and retry the resizing operation later in the next loop.
+
+
+### Risks and Mitigations
+
+- All in-tree volume plugins that support offline file system expansion support online file system expansion:
+  - GCE PD
+  - Azure Disk
+  - AWS EBS
+  - RBD
+  - Cinder
+
+However, Azure Disk in particular does not support online "device" expansion, so in practice a pod needs to be restarted and the device reattached for resize to happen regardless of whether this feature is enabled. File system expansion follows "device"/"controller"/"cloud provider" expansion (i.e. file system expansion is the resize2fs call to enlarge a file system to fill a device, and device expansion is the API call to expand the device in the first place). This difference is surfaced to the user via the "FileSystemResizePending" and "Resizing" conditions, for when file system or cloud provider expansion are in progress respectively. No additional conditions or configuration options will be surfaced to account for this case of online device expansion being unsupported: it is left to the volume plugin implementor & cloud provider to document it and return meaningful errors, which are displayed as events on PVCs.
+
+- Offline resize was introduced first and requires a pod restart to take effect, so online resize becoming the default and occurring immediately upon a PVC edit without requiring a restart may come as a surprise to users. But expansion of mounted file systems is supported by the kernel, so all applications will tolerate it. Plus, if there are cases where users would prefer their volumes to not resize immediately, they can still force offline resize to happen by stopping their pod before editing the PVC, then restarting it. Online resizing will be made default.
+
+- Enabling support of volume expansion through stateful sets (https://github.com/kubernetes/enhancements/issues/661) is orthogonal to making online resizing default. Stateful apps/workloads may need to restart to detect file systems' size changes but should have no issue with online resize.
+
+## Test Plan
+
+There will be e2e tests for online resizing. To isolate vs offline resize, the test will edit the PVC while the Pod is started instead of before.
+
+## Graduation Criteria
+
+### Alpha to Beta
+
+- e2e test where the PVC is edited while the Pod is started instead of before
+- update existing e2e tests so that they test offline resize correctly whether the online resize feature is enabled or not.
```

Comment:
do we have metrics for the online resize operation?

Reply:
yes, with operation name "volume_fs_resize". on the other hand, offline resize (before mount) does not have its own metric I suppose because it's embedded in the mountVolume operation

Reply:
we should probably rename that `volume_node_expand` which would more accurately reflect the intent.  Also since this feature was in alpha before, renaming the metric should not be considered a "breaking" change.

Reply:
controller expand is "expand_volume", so maybe something like "expand_volume_node" to align better?

Reply:
or online resize?

Reply:
I'm ok with that too.

This doc, code, log messages, etc. are guilty of interchanging 'resize' & 'expand' (& 'grow', sometimes). Also 'online' & 'in-use'. The resizefs manpage says 'online resizing' somewhere for what it's worth

Reply:
wait.. we are talking about operation name with which metrics for online resizing should be exposed as. Are you saying `online_resize` should be the operation_name of the metrics?

Reply:
yes i'm referring to the operation metrics name.  Or maybe "online fs expansion" if we want to be consistent with the controller operation name

Reply:
I have been trying to move away from `fs expansion` wording because for CSI expansion on node could mean more than just fs expansion. so `online_node_expansion` works for me. But `online_fs_expansion` is works too.

Reply:
online node expansion sounds good to me

Reply:
Can you add a note for the new metric name somewhere? Maybe this section is fine